src/node_file.cc

```diff
@@ -1150,28 +1153,133 @@ static void RMDir(const FunctionCallbackInfo<Value>& args) {
   }
 }
 
+int MKDirpSync(uv_loop_t* loop, uv_fs_t* req, const std::string& path, int mode,
+               uv_fs_cb cb = nullptr) {
+  FSContinuationData continuation_data = FSContinuationData(req, mode, cb);
+  continuation_data.PushPath(path);
+
+  while (continuation_data.paths.size() > 0) {
+    std::string next_path = continuation_data.PopPath();
+    ssize_t err = uv_fs_mkdir(loop, req, next_path.c_str(), mode, nullptr);
+    switch (err) {
+      case 0:
+        break;
+      case UV_ENOENT: {
+        std::string dirname = next_path.substr(0, path.find_last_of("\\/"));
+        continuation_data.PushPath(next_path);
+        continuation_data.PushPath(dirname);
+        break;
+      }
+      case UV_EPERM: {
+        return err;
+        break;
+      }default:
+        err = uv_fs_lstat(loop, req, next_path.c_str(), nullptr);
+        if (err < 0) return err;
+        if (!S_ISDIR(req->statbuf.st_mode)) return UV_EEXIST;
+        break;
+    }
+  }
+
+  return 0;
+}
+
+int MKDirpAsync(uv_loop_t* loop,
+                uv_fs_t* req,
+                const char* path,
+                int mode,
+                uv_fs_cb cb) {
+  FSReqBase* req_wrap = static_cast<FSReqBase*>(req->data);
```

Comment:
I just landed https://github.com/nodejs/node/pull/21839 – can you use `FSReqBase::from_req(req)` here and below? :)